packages/button/src/LionButton.js

```diff
@@ -213,14 +204,9 @@ export class LionButton extends DisabledWithTabIndexMixin(SlotMixin(LitElement))
    * @param {Event} e
    */
   __clickDelegationHandler(e) {
-    if ((this.type === 'submit' || this.type === 'reset') && e.target === this) {
-      if (this._form) {
-        const nativeButton = document.createElement('button');
-        nativeButton.type = this.type;
-        this._form.appendChild(nativeButton);
-        nativeButton.click();
-        this._form.removeChild(nativeButton);
-      }
+    if (e.target === this) {
+      e.stopImmediatePropagation();
+      this._nativeButtonNode.click();
```

Comment:
The reason why it's not fired from here was the following scenario:

```html
<div @click="${delegatedClickHandler}">
  <lion-button></lion-button>
</div>
```

```js
function delegatedClickHandler(ev) {
  console.log(ev.target); // lion-button, button (2 clicks). 
}

lionButton.click(); // will trigger two clicks

```

In delegatedClickHandler we only want `lion-button` to be found.

Therefore the code we had before should be kept.

TODO: add a unit test for this use case :)

Reply:
The problem of the current way is that you are triggering 2 clicks from one user interaction. It's not delegation, it's duplication. One from `lion-button` and other from the fake button created on the fly. If you listen the click in the form or above you will get 2 events instead of one as expected. Look at what's the current behavior:

![image](https://user-images.githubusercontent.com/948953/95897807-6969f180-0d8e-11eb-893a-88f28ca95daa.png)

I guess is more correct only dispatch one event from the children button because lion-button is on the event path as father. Look at the proposed behavior:

![image](https://user-images.githubusercontent.com/948953/95897636-2740b000-0d8e-11eb-91ad-741b78a4c1ec.png)

If you listen the event in the lion-button you can get it with currentTarget, because the real target who trigger the event is a native button in both cases.

Reply:
I added this test, because the current test 'is fired once' doesn't cover this case:

`
    it('is fired one inside a form', async () => {
      const clickSpy = /** @type {EventListener} */ (sinon.spy());
      const el = /** @type {LionButton} */ (await fixture(
        html`<form @click="${clickSpy}">
          <lion-button>foo</lion-button>
        </form>`,
      ));

      el.click();

      // trying to wait for other possible redispatched events
      await aTimeout(0);
      await aTimeout(0);

      expect(clickSpy).to.have.been.calledOnce
    });
`

Reply:
That is indeed true, but:
- the currentTarget would not work for listeners using event delegation (so in case you don't attach handler to lion-button)
- it creates drawbacks for more commonly used scenarios (see below)
- we would advise to put click listeners (if needed on a form) on ing-form

I think we had your proposed approach in the past, but we decided to go for allowing event delegation for parent elements of lion-button that are in close proximity. Not just for **submit** actions, but mainly for regular clicks on lion-buttons with **type 'button'**.

So we want to support scenarios like these:

```
<consumed-component @click="${onClick}">
    #shadow-root
    <lion-button id="accept">accept</lion-button>
    <lion-button id="cancel">cancel</lion-button>
</consumed-component>
```

As consumer of consumed-component, I'm allowed to know what buttons exist and what ids they have. But I cannot directly add click handlers to those buttons, I need to do it on the host:

```js
function onClick(ev) {
  if (ev.target.id === 'accept') {
    proceed();
  }
}
```
A scenario like this is more likely to happen then that of listening on the form.

But you still have a valid point that on submit we generate 'noise' for the form element (and all its parents) that listen for clicks. We should actually reduce that nosie to a minimum and make sure that propagation is stopped on the form.

Then you got the alternative, not firing a click, but doing `form.submit()`. This creates other drawbacks, @jorenbroekema might know the details about this.

Reply:
### Long story short :)

In case of event delegation you are usually looking for something (like an id). Many events pass, but you filter the things you need (the rest is 'noise').

By using the logic we had, we only introduce a bit (very minimal) extra noise and only on reset/submit and on a native form.

This event would not be catched by the selection criterium of the one creating the event (delegated) listener, because it would not match his/her filter criteria.

Reply:
What about this approaach?
```
      it('has host in the target property', async () => {
        const el = /** @type {LionButton} */ (await fixture('<lion-button type="button">foo</lion-button>'));
        const event = await prepareClickEvent(el);
        expect(event.target).to.equal(el);
      });

      it('has native button in the target property on submit or reset', async () => {
        const el = /** @type {LionButton} */ (await fixture('<lion-button>foo</lion-button>'));
        const event = await prepareClickEvent(el);
        expect(event.target).to.equal(el._nativeButtonNode);
      });
```

In your example the button type should be `button` instead of the default `submit`, then I suggest:

```
<consumed-component @click="${onClick}">
    #shadow-root
    <lion-button type="button" id="accept">accept</lion-button>
    <lion-button type="button" id="cancel">cancel</lion-button>
</consumed-component>
```

In this case it will work as spected

Reply:
Unfortunately the 'problem' with a native button (which we mimic) is that it has type submit by default.
As long as your button is not inside a form, it acts as a regular button (type button).

I think we should mimic that behavior, since most people would not use type=button in the above scenario (and it would be dom that the one applying onClick does not own).

But I agree that it would be nicest to not have the click on the form at all and that your test (the one below) would pass indeed.

```js
it('is fired one inside a form', async () => {
      const clickSpy = /** @type {EventListener} */ (sinon.spy());
      const el = /** @type {HTMLFormElement} */ (await fixture(
        html`<form @click="${clickSpy}">
          <lion-button>foo</lion-button>
        </form>`,
      ));

      // @ts-ignore
      el.querySelector('lion-button').click();

      // trying to wait for other possible redispatched events
      await aTimeout(0);
      await aTimeout(0);

      expect(clickSpy).to.have.been.calledOnce;
    });
```

So maybe we could try to see if we could solve without that extra native button click?
- since we have `this._form`we can try to call `submit()` instead (but I think @jorenbroekema tried before and there were some pitfalls. If I'm correct we listen to the submit event of the native form in lion-form and that wouldn't work anymore)
- we also tried to intercept the event from the native button and adjust the target to lion-button. Although this also led to issues in the end.

Reply:
About the test, currently in master is not passing this test, because lion-button create a native button on the fly and throw the event, then the form and above receive 2 events (fake button and lion-button).

Natural behavior of a `<button type="submit">` inside a form dispatch the button click on submit (by clicking in the button and by pressing enter in a input field).

I tried to simulate `this._form.submit()`, but it send the fom, you can't preventDefault in the form listener.

Change the target it's not allowed, it's a readonly property.

I added `&& this._form` in the condition to check if the button is inside a form to trigger the native click. Otherwise trigger the `lion-button` click.

Reply:
> I added `&& this._form` in the condition to check if the button is inside a form to trigger the native click. Otherwise trigger the `lion-button` click.

Sounds like a very reasonable tradeoff 👍  In that case it needs to act as a 'real submit button' only if part of a form.
Maybe we should then add some tests for this to make the difference (with and without form) clear?

Reply:
The tests are already added, check if they are enough

Reply:
Ok thanks, could you make sure the previous use case is non breaking in case of a regular button (type button)?

```js
it('has host in the target property on `submit` or `reset` inside a form', async () => {
        const el = /** @type {LionButton} */ (await fixture('<lion-button>foo</lion-button>'));
        const clickSpy = /** @type {EventListener} */ (sinon.spy(e => e.preventDefault()));
        await fixture(html`<div @click="${clickSpy}">${el}</div>`);

        const event = await prepareClickEvent(el);

        expect(event.target).to.equal(el);
});
```

I think we can then wrap it up :) Thanks so far 💪

Reply:
I automated all use cases:

```
      const useCases = [
        { container: 'div', type: 'submit', targetHost: true },
        { container: 'div', type: 'reset', targetHost: true },
        { container: 'div', type: 'button', targetHost: true },
        { container: 'form', type: 'submit', targetHost: false },
        { container: 'form', type: 'reset', targetHost: false },
        { container: 'form', type: 'button', targetHost: true },
      ];
```

Feedback is welcome 👍

Reply:
awesome, thanks 💪